feat: Add email input dialog to /booking/uid page

iharsh02 opened the pull request:
## What does this PR do?

Follow-up PR - #18770 

- Fixes #18717  (GitHub issue number)
- Fixes CAL-5045 (Linear issue number - should be visible at the bottom of the GitHub issue description)

This PR adds an email verification dialog to the `/booking/uid` page that:
- Shows immediately when visiting the page without a valid session or email parameter
- Validates the entered email against the booking's attendees and host
- Only displays booking details after successful email verification
- Ensures `cancelledBy` is always set when cancelling a booking

## Updates since last revision

Addressed review feedback from @kart1ka:
- **Implemented server-side email validation**: Added a new `verifyBookingEmail` tRPC endpoint that validates emails against the database instead of client-side validation
- Removed `async` from `updateSearchParams` function
- Removed `await` from `router.replace` call

**Latest update**: Moved prisma query to proper repository pattern:
- Added `findByUidIncludeUserEmailAndAttendeeEmails` method to `BookingRepository`
- Updated handler to use repository instead of importing prisma directly

Files changed:
- `packages/trpc/server/routers/publicViewer/verifyBookingEmail.schema.ts` (new)
- `packages/trpc/server/routers/publicViewer/verifyBookingEmail.handler.ts` (new)
- `packages/features/bookings/repositories/BookingRepository.ts` (modified)

## Visual Demo (For contributors especially)

https://github.com/user-attachments/assets/3628be77-a3d2-4266-9973-a94713f8d6ad

<img width="1384" height="843" alt="image" src="https://github.com/user-attachments/assets/33c8691f-f0a4-45bd-9f9b-63b2ab7b4d5f" />

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox. - N/A
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. Create a booking as a logged-in user
2. Log out and visit `/booking/{uid}` directly
3. Verify the email verification dialog appears immediately
4. Enter an invalid email - should show error message
5. Enter a valid attendee or host email - should show booking details
6. Test cancellation flow - verify `cancelledBy` is set correctly

## Human Review Checklist

- [ ] Verify the new `findByUidIncludeUserEmailAndAttendeeEmails` repository method returns correct data
- [ ] Verify the `verifyBookingEmail` tRPC endpoint correctly validates emails against booking attendees and host
- [ ] Test loading states during email verification
- [ ] Confirm dialog cannot be dismissed without valid email verification
- [ ] Test the flow for both attendees and hosts

## Checklist

- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have checked if my changes generate no new warnings

---

Link to Devin run: https://app.devin.ai/sessions/733eb255221f498981f687837f05b1bc
Requested by: unknown ()

iharsh02 committed:
feat: Add cancelledBy input to cancellation page

iharsh02 committed:
fix: Update email verification error message

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

anikdhabal committed:
Merge branch 'main' into fix-cal-5045

anikdhabal committed:
Update safe-parse.ts

anikdhabal committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
fix: check attendees and booking owner email

anikdhabal committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
fix: minor ui

anikdhabal committed:
Update CancelBooking.tsx

anikdhabal committed:
Update common.json

kart1ka committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
fix: resolve booking-seats.e2e failing test

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

anikdhabal committed:
Update apps/web/components/booking/CancelBooking.tsx

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

alishaz-polymath reviewed (changes requested):
Great effort, thank you 🙌 

However, the fact that anyone can still see the booker email prior to it being prompted at the cancellation step makes it very less useful.

What I would instead recommend is something like this:
- All the details of the booking should be obfuscated when necessary URL params are missing and no session exists
- It should prompt the user to enter their email, it should check for all attendees (including the host). This also helps in cases where the host is trying to cancel the meeting from their phone, they can do that and it marks the meeting cancelled by the host.

This ensures usefulness of the approach as well as it being accurate (to honesty) at the very least, which is what we want here and adds a layer of safety around PII

cc: @CarinaWolli @anikdhabal for visibility and anything else to add here.

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
feat: obfuscate sensitive user information in booking details

iharsh02 committed:
Merge branch 'main' to fix-cal-5045

iharsh02 committed:
chore

iharsh02 committed:
Merge branch 'main' to fix-cal-5045

iharsh02 committed:
feat : Add email input dialog to /booking/uid page

iharsh02 committed:
fix : booking-seats.e2e test

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
fix : linter and coderabbit issue

iharsh02 commented on apps/web/modules/bookings/views/bookings-single-view.tsx:
```diff
@@ -1133,6 +1231,35 @@
+            </div>
+          </div>
+        </DialogContent>
+      </Dialog>
```
`onOpenChange={() => {}}` keeping the onOpenChange will bring new warnings so keep it as it is

iharsh02 committed:
fix : linting and coderabbit issue

iharsh02 committed:
fix: linting issue

iharsh02 commented:
@CarinaWolli @anikdhabal 
#### Changes done :
* Obfuscated all PII (names, emails) when no session or `searchParams` exist.
* show the dialog immediately when visiting `/booking/uid` if no session and `searchParams` avail
* Always set the `cancelled by`

#### Preview :

https://github.com/user-attachments/assets/22882225-821d-42ae-ac70-f3e52a06f8f7

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
fix : booking page being displayed before the dialog box opens

iharsh02 committed:
Merge branch 'main' into fix-cal-5045

iharsh02 committed:
fix : failing e2e test

iharsh02 commented:
> > @CarinaWolli @anikdhabal
> > #### Changes done :
> > 
> > * Obfuscated all PII (names, emails) when no session or `searchParams` exist.
> > * show the dialog immediately when visiting `/booking/uid` if no session and `searchParams` avail
> > * Always set the `cancelled by`
> > 
> > #### Preview :
> > feat-cal-5045.1.mp4
> 
> As per the comment by @CarinaWolli we should not display the booking page at all and not just obfuscate the booking details. In your video I could see that booking page is still being displayed before the dialog box opens.


https://github.com/user-attachments/assets/e11025ae-a607-49a0-8019-3487d84eec17

alishaz-polymath commented:
Hey @kart1ka can we please prioritize a review here and get this over the line?

kart1ka committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
Merge branch 'main' into fix-cal-5045

kart1ka reviewed (changes requested):
Left a few comments.

kart1ka committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
Merge branch 'main' into fix-cal-5045

kart1ka committed:
Merge branch 'main' into fix-cal-5045

devin-ai-integration[bot] committed:
Merge upstream/main into fix-cal-5045 - resolve conflicts for email verification feature

Co-Authored-By: unknown <>

keithwillcode committed:
Merge branch 'main' into fix-cal-5045

devin-ai-integration[bot] committed:
feat: implement server-side email verification for booking page

- Add verifyBookingEmail tRPC endpoint for server-side email validation
- Replace client-side email validation with server-side validation
- Remove async/await from updateSearchParams as per review feedback
- Address review comment by @kart1ka for server-side validation

Co-authored-by: iharsh02 <iharsh02@users.noreply.github.com>
Co-Authored-By: unknown <>

keithwillcode committed:
Merge branch 'main' into fix-cal-5045

devin-ai-integration[bot] committed:
refactor: move prisma query to BookingRepository

- Add findByUidIncludeUserEmailAndAttendeeEmails method to BookingRepository
- Update verifyBookingEmail handler to use repository instead of prisma directly
- Follow Cal.com's repository pattern for data access

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: address review feedback for email verification feature

- Add .trim() to email validation schema to handle pasted emails with whitespace
- Add try/catch error handling to handleCancellation in CancelBooking.tsx
- Reset isEmailVerified state when email param changes to prevent state persistence bug
- Add loading state to verify button for better UX

Co-authored-by: iharsh02 <iharsh02@users.noreply.github.com>
Co-Authored-By: unknown <>
